File: crates/dashboard-ui/src/lib/artifactKind.ts
export type ArtifactKind =
  | "image"
  | "video"
  | "audio"
  | "pdf"
  | "presentation"
  | "document"
  | "mindmap"
  | "report"
  | "file"
  | "media";

const IMAGE_EXT = new Set(["png", "jpg", "jpeg", "gif", "webp", "svg", "bmp", "ico", "avif"]);
const VIDEO_EXT = new Set(["mp4", "mov", "webm", "mkv", "m4v", "avi"]);
const AUDIO_EXT = new Set(["mp3", "wav", "ogg", "m4a", "aac", "flac"]);
const PRESENTATION_EXT = new Set(["pptx", "ppt", "key", "odp"]);
const DOCUMENT_EXT = new Set(["docx", "doc", "xlsx", "xls", "csv", "txt", "rtf", "odt", "ods"]);
const REPORT_EXT = new Set(["html", "htm"]);
const MINDMAP_EXT = new Set(["mm", "xmind"]);

function extension(path: string): string {
  const base = path.split(/[/\\]/).pop() ?? path;
  const dot = base.lastIndexOf(".");
  if (dot <= 0) return "";
  return base.slice(dot + 1).toLowerCase();
}

function normalizeKind(value: string | undefined | null): ArtifactKind | null {
  if (!value) return null;
  const lower = value.toLowerCase();
  const kinds: ArtifactKind[] = [
    "image",
    "video",
    "audio",
    "pdf",
    "presentation",
    "document",
    "mindmap",
    "report",
    "file",
    "media",
  ];
  if (kinds.includes(lower as ArtifactKind)) {
    return lower as ArtifactKind;
  }
  if (lower.includes("presentation") || lower.includes("slideshow")) return "presentation";
  if (lower.includes("document")) return "document";
  if (lower.includes("mindmap") || lower.includes("mind-map")) return "mindmap";
  if (lower.includes("report")) return "report";
  if (lower.includes("image")) return "image";
  if (lower.includes("video")) return "video";
  if (lower.includes("audio")) return "audio";
  if (lower.includes("media")) return "media";
  return null;
}

export function kindForPath(path: string, hint?: string | null): ArtifactKind {
  const fromHint = normalizeKind(hint);
  if (fromHint && fromHint !== "file" && fromHint !== "media") {
    return fromHint;
  }

  const ext = extension(path);
  const lowerPath = path.toLowerCase();

  if (ext === "pdf") return "pdf";
  if (IMAGE_EXT.has(ext)) return "image";
  if (VIDEO_EXT.has(ext)) return "video";
  if (AUDIO_EXT.has(ext)) return "audio";
  if (PRESENTATION_EXT.has(ext)) return "presentation";
  if (MINDMAP_EXT.has(ext) || lowerPath.includes("mindmap") || lowerPath.includes("mind-map")) {
    return "mindmap";
  }
  if (ext === "md" && (lowerPath.includes("mindmap") || lowerPath.includes("mind-map"))) {
    return "mindmap";
  }
  if (REPORT_EXT.has(ext)) return "report";
  if (DOCUMENT_EXT.has(ext) || ext === "md") return "document";

  if (fromHint) return fromHint;
  return "file";
}

export function mimeForPath(path: string, hint?: string | null): string {
  if (hint?.includes("/")) return hint;

  const ext = extension(path);
  const map: Record<string, string> = {
    png: "image/png",
    jpg: "image/jpeg",
    jpeg: "image/jpeg",
    gif: "image/gif",
    webp: "image/webp",
    svg: "image/svg+xml",
    bmp: "image/bmp",
    avif: "image/avif",
    mp4: "video/mp4",
    mov: "video/quicktime",
    webm: "video/webm",
    mkv: "video/x-matroska",
    mp3: "audio/mpeg",
    wav: "audio/wav",
    ogg: "audio/ogg",
    m4a: "audio/mp4",
    pdf: "application/pdf",
    pptx: "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    ppt: "application/vnd.ms-powerpoint",
    docx: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    doc: "application/msword",
    xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    xls: "application/vnd.ms-excel",
    csv: "text/csv",
    txt: "text/plain",
    md: "text/markdown",
    html: "text/html",
    htm: "text/html",
    json: "application/json",
  };
  return map[ext] ?? "application/octet-stream";
}

export function isInlineKind(kind: ArtifactKind): boolean {
  return (
    kind === "image" ||
    kind === "video" ||
    kind === "audio" ||
    kind === "mindmap" ||
    kind === "pdf" ||
    kind === "presentation" ||
    kind === "document" ||
    kind === "media"
  );
}

File: crates/core/src/artifact_kind.rs
//! Artifact kind / MIME mapping (single source for runtime + dashboard heuristics).

use std::path::Path;

/// Canonical artifact kinds used in DB `artifacts.kind` and transcript cards.
pub fn artifact_kind_for_path(path: &str) -> &'static str {
    let lower = path.to_lowercase();
    let ext = Path::new(&lower)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("");
    match ext {
        "png" | "jpg" | "jpeg" | "gif" | "webp" | "svg" | "bmp" | "heic" => "image",
        "mp4" | "webm" | "mov" | "m4v" | "mkv" => "video",
        "mp3" | "wav" | "m4a" | "aac" | "ogg" | "flac" => "audio",
        "pdf" => "pdf",
        "pptx" | "ppt" | "key" => "presentation",
        "docx" | "doc" | "rtf" | "odt" => "document",
        "xlsx" | "xls" | "csv" | "ods" => "spreadsheet",
        "md" | "markdown" => {
            if lower.contains("mindmap") || lower.contains("mind-map") || lower.contains("导图") {
                "mindmap"
            } else if lower.contains("report") {
                "report"
            } else {
                "file"
            }
        }
        "mmd" => "mindmap",
        "json" if lower.contains("workbook") => "spreadsheet",
        // Any HTML is a previewable report card (align with dashboard-ui kindForPath).
        "html" | "htm" => "report",
        "ipynb" => "notebook",
        _ => "file",
    }
}

/// Guess MIME from path extension.
pub fn mime_for_path(path: &str) -> &'static str {
    let lower = path.to_lowercase();
    let ext = Path::new(&lower)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("");
    match ext {
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "gif" => "image/gif",
        "webp" => "image/webp",
        "svg" => "image/svg+xml",
        "bmp" => "image/bmp",
        "mp4" => "video/mp4",
        "webm" => "video/webm",
        "mov" => "video/quicktime",
        "mkv" => "video/x-matroska",
        "mp3" => "audio/mpeg",
        "wav" => "audio/wav",
        "m4a" => "audio/mp4",
        "ogg" => "audio/ogg",
        "pdf" => "application/pdf",
        "pptx" => "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "docx" => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "xlsx" => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "csv" => "text/csv",
        "md" | "markdown" | "mmd" => "text/markdown",
        "json" => "application/json",
        "html" | "htm" => "text/html",
        _ => "application/octet-stream",
    }
}

/// Whether this kind should render an inline card in the conversation stream.
pub fn artifact_kind_is_inline(kind: &str) -> bool {
    matches!(
        kind,
        "image"
            | "video"
            | "audio"
            | "pdf"
            | "mindmap"
            | "presentation"
            | "document"
            | "spreadsheet"
            | "media"
            | "report"
    )
}

/// Title from file name.
pub fn artifact_title_for_path(path: &str) -> String {
    Path::new(path)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or(path)
        .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn kinds_and_mime() {
        assert_eq!(artifact_kind_for_path("/a/b/c.png"), "image");
        assert_eq!(mime_for_path("/a/b/c.png"), "image/png");
        assert_eq!(artifact_kind_for_path("deck.pptx"), "presentation");
        assert_eq!(artifact_kind_for_path("notes.mindmap.md"), "mindmap");
        assert_eq!(artifact_kind_for_path("sheet.xlsx"), "spreadsheet");
        assert_eq!(artifact_kind_for_path("data.json"), "file");
        assert_eq!(artifact_kind_for_path("sales-workbook.json"), "spreadsheet");
        assert_eq!(artifact_kind_for_path("deck.html"), "report");
        assert_eq!(artifact_kind_for_path("notes.md"), "file");
        assert!(artifact_kind_is_inline("spreadsheet"));
        assert!(artifact_kind_is_inline("image"));
        assert!(artifact_kind_is_inline("report"));
        assert!(!artifact_kind_is_inline("bash"));
    }
}
